File: src/main/java/ru/ifmo/mit/repl/command/CommandFactoryImpl.java
package ru.ifmo.mit.repl.command;

import ru.ifmo.mit.repl.env.Context;

import java.util.List;

/*
    Ипмлементация фабрики для создания команд шелла.
 */
public class CommandFactoryImpl extends CommandFactory {
    private final Context context;

    public CommandFactoryImpl(Context context) {
        this.context = context;
    }

    @Override
    public CommandExecutable makeAssignCommand(List<String> arguments) {
        return new CommandAssign(arguments, context);
    }

    @Override
    public CommandExecutable makeCatCommand(List<String> arguments) {
        return new CommandCat(arguments);
    }

    @Override
    public CommandExecutable makeEchoCommand(List<String> arguments) {
        return new CommandEcho(arguments);
    }

    @Override
    public CommandExecutable makeExternalCommand(List<String> arguments) {
        return new CommandExternal(arguments);
    }

    @Override
    public CommandExecutable makePipeCommand(List<CommandExecutable> commands) {
        return new CommandPipe(commands);
    }

    @Override
    public CommandExecutable makePwdCommand(List<String> arguments) {
        return new CommandPWD(arguments);
    }

    @Override
    public CommandExecutable makeWCCommand(List<String> arguments) {
        return new CommandWC(arguments);
    }

    @Override
    public CommandExecutable makeExitCommand(List<String> arguments) { return new CommandExit(arguments); }

    @Override
    public CommandExecutable makeGrepCommand(List<String> arguments) { return new CommandGrep(arguments); }

    @Override
    public CommandExecutable makeLsCommand(List<String> arguments) { return new CommandLs(arguments); }

    @Override
    public CommandExecutable makeCdCommand(List<String> arguments) { return new CommandCd(arguments); }
}

File: src/main/kotlin/command/CommandFactoryImpl.kt
package command

import executor.ShellContext
import parse.*

class CommandFactoryImpl(private val context: ShellContext) : CommandFactory() {
    override fun makeAssignCommand(assign: Assign): ShellCommand = AssignCommand(assign, context)

    override fun makeCatCommand(cat: Cat): ShellCommand = CatCommand(cat)

    override fun makeEchoCommand(echo: Echo): ShellCommand = EchoCommand(echo)

    override fun makeExternalCommand(external: External): ShellCommand = ExternalCommand(external)

    override fun makePipeCommand(pipe: Pipe): ShellCommand = PipeCommand(pipe.commands.map(this::makeCommand))

    override fun makePwdCommand(pwd: Pwd): ShellCommand = PwdCommand()

    override fun makeWCCommand(wc: WC): ShellCommand = WCCommand(wc)

    override fun makeExitCommand(exit: Exit): ShellCommand = ExitCommand()

    override fun makeGrepCommand(grep: Grep): ShellCommand = GrepCommand(grep)
}
